django/db/models/expressions.py

```diff
@@ -519,7 +519,7 @@ def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize
             c.source_expressions[pos] = arg.resolve_expression(query, allow_joins, reuse, summarize, for_save)
         return c
 
-    def as_sql(self, compiler, connection, function=None, template=None):
+    def as_sql(self, compiler, connection, **kwargs):
```

Comment:
Ok, let's change this to:

```
def as_sql(self, compiler, connection, function=None, template=None, arg_joiner=None, **extra_context):
```

You'll need to conditionally add function to extra_context before merging it with self.extra:

```
if function is not None:
    extra_context['function'] = function
arg_joiner = arg_joiner or self.arg_joiner
template = template or self.template
```

etcetera.

The docs for `Func` will need to call out the extra key word arguments for as_sql specifically.